specs/phase1/beacon-chain.md

```diff
@@ -768,234 +830,264 @@ def validate_attestation(state: BeaconState, attestation: Attestation) -> None:
-                    pending_attestation.crosslink_success = True
-```
+def process_shard_transition(state: BeaconState, transition: ShardTransition) -> None:
+    # TODO: only need to check it once when phase 1 starts
+    assert state.slot > PHASE_1_FORK_SLOT
+    # Do not attempt to transition the Genesis transition stub
+    assert transition != ShardTransition()
+
+    # Extract matching ShardTransitionCandidate
+    all_candidates = state.current_shard_transition_candidates + state.previous_shard_transition_candidates
+    latest_block_root = transition.shard_states[len(transition.shard_states) - 1].latest_block_root
+    matching_candidates = [
+        c for c in all_candidates
+        if (c.data.transition_root, c.data.block_root, get_shard_from_transition_candidate(state, c), c.data.slot)
+        == (hash_tree_root(transition), latest_block_root, transition.shard, transition.committee_slot)
+    ]
+    assert len(matching_candidates) == 1
+    matching_candidate = matching_candidates[0]
 
-###### `verify_empty_shard_transition`
+    # Validate the shard transition
+    validate_shard_transition(state, transition, matching_candidate)
 
-```python
-def verify_empty_shard_transition(state: BeaconState, shard_transitions: Sequence[ShardTransition]) -> bool:
-    """
-    Verify that a `shard_transition` in a block is empty if an attestation was not processed for it.
-    """
-    for shard in range(get_active_shard_count(state)):
-        if state.shard_states[shard].slot != compute_previous_slot(state.slot):
-            if shard_transitions[shard] != ShardTransition():
-                return False
-    return True
+    # Apply the transition updates
+    apply_shard_transition_updates(state, transition, matching_candidate)
 ```
 
-###### `process_shard_transitions`
+##### Verify ShardTransitions
+
+###### `validate_shard_transitions`
 
 ```python
-def process_shard_transitions(state: BeaconState,
-                              shard_transitions: Sequence[ShardTransition],
-                              attestations: Sequence[Attestation]) -> None:
-    # NOTE: We currently set `PHASE_1_FORK_SLOT` to `GENESIS_SLOT` for test vectors.
-    if compute_previous_slot(state.slot) > GENESIS_SLOT:
-        # Process crosslinks
-        process_crosslinks(state, shard_transitions, attestations)
-
-    # Verify the empty proposal shard states
-    assert verify_empty_shard_transition(state, shard_transitions)
+def validate_shard_transitions(state: BeaconState, shard_transitions: Sequence[ShardTransition]) -> None:
+    # Validate shard transitions in sorted order
+    assert shard_transitions == sorted(shard_transitions, key=lambda transition: transition.shard)
+
+    # Validate shard transition uniqueness
+    for shard in range(get_active_shard_count(state)):
+        transitions_for_shard = [transition for transition in shard_transitions if transition.shard == shard]
+        assert len(transitions_for_shard) <= 1
```

Comment:
Are we not allowing multiple `shard_transition` under the same shard? What if there's 2 `shard_transition` representing 2 forked chains and both have equal weights.